Fix Matrix server sorting

ulope opened the pull request:
## Description

Fixes: #5626

In #5544 the algorithm to sort the available Matrix servers by response time was changed to only consider the first result.

As described in #5626, this caused some issues, namely:
- If a request failed (e.g. name resolution, network unreachable) that greenlet would return first, causing a false positive.
- The `make_client()` function that is using `sort_servers_closest()` has the option of falling back to a less optimal server in case the 'best' one (by response time) turned out to be unusable for some other reason. This was no longer possible with only one result.

This changes the algorithm back to return multiple results but still limit the worst case time we wait to gather those to a reasonable amount (7s currently), so that slow servers can't arbitrarily delay node startup, which was the initial motivation for #5544.

hackaugusto commented on raiden/network/transport/matrix/utils.py:
```diff
@@ -637,35 +637,40 @@ def validate_userid_signature(user: User) -> Optional[Address]:
-    result = gevent.wait(rtt_greenlets, count=1)[0].get()
+    results = []
+    # `get_average_http_response_time` performs 3 requests with 0.125 s delay inbetween
+    for greenlet in gevent.iwait(rtt_greenlets, timeout=max_timeout * 3 + 1):
```
This will iterate over all servers, the result will that server selection will be as fast as the slowest server.

hackaugusto replied:
Please, don't have multiple timeouts for the same thing ... this is just hard to debug and compose. The timeout here is not necessary since the `get_average_http_response_time` has a timeout already, and it will return `None` if the request itself timedout.

ulope commented:
> that server selection will be as fast as the slowest server.

Yes, bounded by the timeout. That doesn't seem so bad to me for something only done once on startup.

> The timeout here is not necessary since the `get_average_http_response_time` has a timeout already

Unfortunately that is only a requests / socket timeout which is vulnerable to slow read (e.g. 1 byte every 1.9 seconds with the current 2s timeout will not trigger the timeout).
But you're right stacking timeouts is very annoying.
I guess switching to a gevent timeout inside `get_average_http_response_time` and removing the timeout here would be a better solution.

hackaugusto commented:
> Yes, bounded by the timeout. That doesn't seem so bad to me for something only done once on startup.

I changed it specifically because the startup is too slow.

> Unfortunately that is only a requests / socket timeout which is vulnerable to slow read (e.g. 1 byte every 1.9 seconds with the current 2s timeout will not trigger the timeout).

A MTU has to be at least 68 bytes, the HEAD request is 208 (as specified in the `Content-length`), the header is 153 (I measured it with a curl), that means at *maximum* 4 packets to get the full response ... which is pretty much the second timeout because of the `*3`. On top of that, the default value for the MTU is 1500, which means a single packet.

Again, timeouts don't compose, makes things harder to understand, and in this case it is not achieving what it wants.

ulope commented:
> I changed it specifically because the startup is too slow.

Yes, I know. But in the old behaviour it would wait indefinitely for slow servers.
Do you think 6s (or 9s with the change to a 3s pre-request timeout) is unreasonable for the *worst case*?

> A MTU has to be at least 68 bytes, the HEAD request is 208 (as specified in the Content-length), the header is 153 (I measured it with a curl), that means at maximum 4 packets to get the full response ... which is pretty much the second timeout because of the *3. On top of that, the default value for the MTU is 1500, which means a single packet.

You're forgetting TLS overhead in this calculation. Establishing a new TLS session requires 4 - 6kB according to this: https://stackoverflow.com/questions/1615882/how-much-network-overhead-does-tls-add-compared-to-a-non-encrypted-connection. 

My point was that the requests based timeout isn't reliable to guarantee that will not encounter a network situation that could cause long hangs.

> Again, timeouts don't compose, makes things harder to understand, and in this case it is not achieving what it wants.

Yes, I agree. I only see the option of removing the timeout from one of the two places.

hackaugusto commented:
> Yes, I know. But in the old behaviour it would wait indefinitely for slow servers.

True. We could  just have a `timeout` in the `wait` in addition to the `count`.

> My point was that the requests based timeout isn't reliable to guarantee that will not encounter a network situation that could cause long hangs.

Lets remove the read timeout then, and instead use a `gevent.Timeout` for wall-time timeout.

ulope commented:
Even though there was a requests timeout in the old version (before your change) we still had very long hangs. 

The PR is now changed as discussed above.
I implemented the timeout in `sort_servers_closest` instead of `get_average_http_response_time` because having `Timeout`s in multiple parallel greenlets behaved very strangely in the tests and this seems cleaner anyway.
